File: src/main/java/fi/quanfoxes/parser/nodes/VariableNode.java
package fi.quanfoxes.parser.nodes;

import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Contextable;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Variable;

public class VariableNode extends Node implements Contextable {
    private Variable variable;

    public VariableNode(Variable variable) {
        this.variable = variable;
        this.variable.addUsage(this);
    }

    public Variable getVariable() {
        return variable;
    }

    @Override
    public Context getContext() throws Exception {
        return variable.getType();
    }
    
    @Override
    public NodeType getNodeType() {
        return NodeType.VARIABLE_NODE;
    }
}

File: Zigzag/Parser/Nodes/VariableNode.cs
using System;
using System.Collections.Generic;

public class VariableNode : Node, IType
{
	public Variable Variable { get; set; }

	public VariableNode(Variable variable)
	{
		Variable = variable;
		Variable.References.Add(this);
	}

	public new Type? GetType()
	{
		return Variable.Type;
	}

	public override NodeType GetNodeType()
	{
		return NodeType.VARIABLE_NODE;
	}

	public override bool Equals(object? obj)
	{
		return obj is VariableNode node &&
				base.Equals(obj) &&
				EqualityComparer<Variable>.Default.Equals(Variable, node.Variable);
	}

	public override int GetHashCode()
	{
		return HashCode.Combine(base.GetHashCode(), Variable);
	}
}
